containers/daap-python-base/src/var/task/data_product_metadata.py

```diff
@@ -0,0 +1,117 @@
+        "metadata",
+        data_product_name,
+        "v1.0",
+        "metadata.json"
+    )
+    return metadata_s3_path
+
+
+def get_data_product_metadata_spec_path(version: str = None) -> str:
+    # if version is None we'll get the latest version
+    if version is None:
+        file_paths = get_filepaths_from_s3_folder(
+            os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec")
+        )
+        versions = list(
+            {i for p in file_paths for i in p.split("/") if i.startswith("v")}
+        )
+        versions.sort()
+        latest_version = versions[-1]
+        spec_path = os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec", latest_version, "moj_data_product_metadata_spec.json")
+    else:
+        spec_path = os.path.join("s3://", get_bucket_name(), "data_product_metadata_spec", version, "moj_data_product_metadata_spec.json")
+
+    return spec_path
+
+
+def split_bucket_and_key(path):
+    bucket, key = path.replace("s3://", "").split("/", 1)
+    return bucket, key
+
+
+class DataProductMetadata():
+    """
+    class to handle creation and updating of
+    metadata relating to a dataproduct
+    """
+    def __init__(self, data_product_name: str, logger: DataPlatformLogger):
+        self.data_product_name = data_product_name
+        self.logger = logger
+        bucket, key = split_bucket_and_key(get_data_product_metadata_path(data_product_name))
+        self.metadata_bucket = bucket
+        self.metadata_key = key
+        self._check_if_metadata_exists()
+
+
+    def _check_if_metadata_exists(self) -> bool:
+        # establish whether metadata for data product already exists
+        try:
+            # get head of object (if it exists)
+            s3_client.head_object(Bucket=self.metadata_bucket, Key=self.metadata_key)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                self.metadata_exists = False
+            else:
+                self.logger.error(f"Uknown error - {e}")
+                raise Exception(f"Uknown error - {e}")
+        else:
+            self.logger.info("version 1 of metadata already exists for this data product")
+            self.metadata_exists = True
+
+        return self
+
+    def validate_metadata(
+        self,
+        data_product_metadata: dict,
+        metadata_schema_version: str = None,
+    ) -> bool:
+
+        metadata_schema = read_json_from_s3(get_data_product_metadata_spec_path(metadata_schema_version))
+        try:
+            validate(instance=data_product_metadata, schema=metadata_schema)
+        except ValidationError:
+            self.error_traceback = traceback.format_exc()
+            self.logger.error(f"metadata has failed validation with error: {self.error_traceback}")
+            self.valid_metadata = False
+        else:
+            self.logger.info("metadata has passed validation")
+            self.valid_metadata = True
```

Comment:
Is this object usable by calling code before you've passed in the `data_product_metadata` and `metadata_schema_version` to the validate method? If not, you could move these to the initializer.

I think it might also streamline the calling code if the validate function returns the boolean directly instead of `self`.

I'm thinking something along the lines of this:

```python
metadata = DataProductMetadata(data_product_name, data_product_metadata, metadata_schema_version)
if not metadata.validate():
  # return error response

metadata.write_to_s3()
```

Reply:
My thinking was to expand this class to not only create the first data product metadata file, but subsequent updates to metadata. So need to test whether the v1 of the metadata exists already because if it does then a different endpoint should be used (but it'll use this class too) and so validation is not necessary if a version of the data product metadata already exists